### PR TITLE
Refactor `TransferPackage` to transfer boxes instead of samples

### PR DESCRIPTION
**Pull request by straight-shoota**
Ref https://github.com/instedd/cdx/issues/1610

grafik

### CODE REVIEW

**Comment by straight-shoota** [comment]
I have implemented the Box transfer feature as a replacement of the individual sample transfers. 
I'm not sure if individual sample transfers are in use in production. If not, we could probably just drop them entirely. For now, I have kept the old sample transfers index so you can see existing transfers. But individual sample transfers can't be created nor confirmed anymore because this would conflict with the Box transfer model. Eventually, we want to migrate individual sample transfers to box transfers, but this depends on #1619. So we're temporarily losing the capability to transfer individual samples. This is caused by prioritizing box transfers for already existing purpose boxes.

**Comment by diegoliberman** [comment]
@straight-shoota the individual sample transfers that were done in Production, were just tests.  We could write a migration to delete them, and probably that's the easiest and best solution.

**Review by ysbaddaden — APPROVED** [review]
The code looks good. I haven't tested it, yet.

File: app/models/box.rb [hunk 1]
```diff
@@ -47,4 +63,22 @@ def build_samples(batch, concentration_exponents:, replicates:, media:)
...
+
+  def detach_from_context
+    assign_attributes(
+      site: nil,
+      institution: nil
+    )
+    samples.each do |sample|
+      sample.detach_from_context
+      sample.save!
+    end
```
[the remark this answers is not in the document]
Comment (straight-shoota) [reply app/models/box.rb:82]: Good catch! 😓
I actually implemented the other end as a single query (for the confirm action) but forgot about this one.

[the remark this answers is not in the document]
Comment (straight-shoota) [reply app/models/box.rb:82]: Urgh, it's actually not trivial to implement the assignment to `old_batch_number` 😞

File: app/models/transfer_package.rb [hunk 2]
```diff
@@ -30,11 +33,11 @@ class TransferPackage < ApplicationRecord
...
   before_create do
-    sample_transfers.each do |sample_transfer|
-      sample = sample_transfer.sample
-      sample.attach_qc_info if includes_qc_info
-      sample.detach_from_context unless confirmed?
-      sample.save!
+    box_transfers.each do |box_transfer|
+      box = box_transfer.box
+      box.attach_qc_info if includes_qc_info
+      box.detach_from_context unless confirmed?
```
[the remark this answers is not in the document]
Comment (straight-shoota) [reply app/models/transfer_package.rb:39]: That's what I thought 😆

File: app/models/box.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (ysbaddaden) [reply app/models/box.rb]: Ooooh, I skipped that. Hum, maybe:

```ruby
samples
  .left_joins(:batch)
  .update_all("samples.old_batch_number = batches.batch_number, institution_id = NULL, site_id = NULL, batch_id = NULL")
```

[the remark this answers is not in the document]
Comment (ysbaddaden) [reply app/models/box.rb]: Except that `#old_batch_number` is an entity field :bomb: 

Let's forget about it.

**Comment by ysbaddaden** [comment]
@straight-shoota the problem is that `#old_batch_number` is an entity field with the `copy: true` option, and it must exist both as a column and inside the `core_fields` JSON, so we can't have an UPDATE with a LEFT JOIN, and the `core_fields` column isn't a JSON column and/or I'm not sure we can alter a JSON column in a SQL request with MySQL 5.7 (of course with PG we could) :sob:

**Comment by ysbaddaden** [comment]
**AND** it overwrites the `#old_batch_number` to `nil` when we re-transfer a previously transferred sample :upside_down_face: thought it could be fixed with `COALESCE(old_batch_number, batches.batch_number)`.

Oh it looks like we could use [JSON_SET()](https://dev.mysql.com/doc/refman/5.7/en/json-modification-functions.html#function_json-set).

**Comment by straight-shoota** [comment]
But well anyways, this effectively means we'll have to go back to update every single sample individually, right? At least for now.

**Comment by ysbaddaden** [comment]
That's the entity framework for you. I'm not really sure how it works, but most fields are entities first (stored in a JSON) and a column afterwards to be able to make some queries. To be honest I'm not exactly sure how the whole thing works (it seems to be related with the elasticsearch queries).

The `core_fields` and `custom_fields` columns should become a JSON data type, but I'm not sure which AR version supports it. Then we'd be able to use JSON functions and to make queries inside the JSON (granted that we create autogenerated virtual columns to have indexes). We could do a very nice refactor of the entity framework (we should define the core fields inside the models + use plain columns for them, use the YAML definition for custom, per-project, fields, then auto-generate virtual columns, indexes and scopes), but we don't have time to deal with that right now.

**Comment by straight-shoota** [comment]
Sounds fun 🤦

**Pull request merged**